htsget/htsget.go

```diff
@@ -0,0 +1,252 @@
+				UserAgent      string `json:"user-agent"`
+				Host           string `json:"host"`
+				AcceptEncoding string `json:"accept-encoding"`
+				Authorization  string `json:"authorization"`
+			} `json:"headers,omitempty"`
+		} `json:"urls"`
+	} `json:"htsget"`
+}
+
+// Htsget function downloads the files included in the urls_list.txt file.
+// The argument can be a local file or a url to an S3 folder
+func Htsget(args []string) error {
+
+	// Call ParseArgs to take care of all the flag parsing
+	err := helpers.ParseArgs(args, Args)
+	if err != nil {
+		return fmt.Errorf("failed parsing arguments, reason: %v", err)
+	}
+
+	if *datasetID == "" || *fileName == "" || *htsgetHost == "" || *publicKeyFile == "" {
+		return fmt.Errorf("missing required arguments, dataset, filename, htsgethost and key are required")
+	}
+
+	config, err := helpers.GetAuth(*configPath)
+	if err != nil {
+		return err
+	}
+
+	// read public key from file
+	publickey, err := os.ReadFile(*publicKeyFile)
+	if err != nil {
+		return fmt.Errorf("failed to read public key, reason: %v", err)
+	}
+	base64publickey := base64.StdEncoding.EncodeToString(publickey)
+
+	// TODO: Add cases for different type of files
+	// i.e. bam files require the /reads/, replace for vcf
+	url := *htsgetHost + "/reads/" + *datasetID + "/" + *fileName
+	if *referenceName != "" {
+		url = url + "?referenceName=" + *referenceName
+	}
+	method := "GET"
+	client := &http.Client{}
+	req, err := http.NewRequest(method, url, nil)
+	if err != nil {
+		return fmt.Errorf("failed to make request, reason: %v", err)
+	}
+
+	req.Header.Add("Authorization", "Bearer "+config.AccessToken)
+	req.Header.Add("client-public-key", base64publickey)
+
+	res, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to do the request, reason: %v", err)
+	}
+	if res.StatusCode != 200 {
+		return fmt.Errorf("failed to get the file, status code: %v", res.StatusCode)
+	}
+	defer res.Body.Close()
+
+	body, err := io.ReadAll(res.Body)
+	if err != nil {
+		return fmt.Errorf("failed to read response, reason: %v", err)
+	}
+
+	var htsgetURLs htsgetResponse
+	err = json.Unmarshal(body, &htsgetURLs)
+	if err != nil {
+		return fmt.Errorf("error unmarshaling response, reason: %v", err)
+	}
+
+	err = downloadFiles(htsgetURLs, config)
+	if err != nil {
+		return fmt.Errorf("error downloading the files, reason: %v", err)
+	}
+
+	return nil
+
+}
+
+func downloadFiles(htsgeURLs htsgetResponse, config *helpers.Config) (err error) {
+
+	// Create the directory for the file
+	var currentPath, filePath, filenameToUse string
+	currentPath, err = os.Getwd()
+	if err != nil {
+		return fmt.Errorf("failed to get current path, reason: %v", err)
+	}
+	if strings.Contains(*fileName, string(os.PathSeparator)) {
+		filePath = filepath.Dir(*fileName)
+		err = os.MkdirAll(filePath, os.ModePerm)
+		if err != nil {
+			return fmt.Errorf("failed to create file path, reason: %v", err)
+		}
+	}
+	filenameToUse = *fileName + ".c4gh"
```

Comment:
htsget could be setup to return decrypted files. In that case it would not be correct to add the `c4gh` extension.

Reply:
Added a check to figure out if the file in the url is encrypted or not.
 If it's encrypted, we add a c4gh extension to the filename, unless defined otherwise with the -output argument
 To test this, run the htsget server with the configuration `send_encrypted_to_client = false` in `download-config.toml`. 
 Then run the example:
 ```
 ./sda-cli htsget -config testing/s3cmd.conf -dataset DATASET0001 -filename htsnexus_test_NA12878 -reference 11 -host http://localhost:8088 -pubkey c4gh.pub.pem
 ```
 This should create the file htsnexus_test_NA12878 without the extension c4gh.

Reply:
Great! Perhaps it's worth adding an issue for getting this info directly from htsget or download, in the future?